en/lessons/basics/functions.md

```diff
@@ -1,425 +1,83 @@
-We'll learn more about Modules in the next lessons, for now we'll focus on the named functions alone.
+And incidentally, we have also seen one of the most primordial functions : The identity function, or `id`.  
+This function is very simple, as it returns precisely the argument that we pass to it.
 
-Functions defined within a module are available to other modules for use.
-This is a particularly useful building block in Elixir:
-
-```elixir
-defmodule Greeter do
-  def hello(name) do
-    "Hello, " <> name
-  end
-end
-
-iex> Greeter.hello("Sean")
-"Hello, Sean"
-```
+## Named functions
 
-If our function body only spans one line, we can shorten it further with `do:`:
+However, one can only go so far with anonymous functions. This is why we can give names to our functions:
 
-```elixir
-defmodule Greeter do
-  def hello(name), do: "Hello, " <> name
-end
-```
-
-Armed with our knowledge of pattern matching, let's explore recursion using named functions:
-
-```elixir
-defmodule Length do
-  def of([]), do: 0
-  def of([_ | tail]), do: 1 + of(tail)
-end
-
-iex> Length.of []
-0
-iex> Length.of [1, 2, 3]
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> addOne x = x + 1
+ghci> addOne 2
 3
 ```
 
-### Function Naming and Arity
+## Higher-order functions
 
-We mentioned earlier that functions are named by the combination of given name and arity (number of arguments).
-This means you can do things like this:
+We can now see a third way of using functions, which is the composition of both anonymous and named functions:
+Higher-order functions!
 
-```elixir
-defmodule Greeter2 do
-  def hello(), do: "Hello, anonymous person!"   # hello/0
-  def hello(name), do: "Hello, " <> name        # hello/1
-  def hello(name1, name2), do: "Hello, #{name1} and #{name2}"
-                                                # hello/2
-end
+They become particularly useful when you need to control the application of a function over a parameter:
 
-iex> Greeter2.hello()
-"Hello, anonymous person!"
-iex> Greeter2.hello("Fred")
-"Hello, Fred"
-iex> Greeter2.hello("Fred", "Jane")
-"Hello, Fred and Jane"
+```console?lang=haskell&prompt=ghci>,ghci|
+ghci> apply2 f x = f (f x)
+ghci> apply2 addOne 2
+4
 ```
 
-We've listed the function names in comments above.
-The first implementation takes no arguments, so it is known as `hello/0`; the second takes one argument so it is known as `hello/1`, and so on.
-Unlike function overloads in some other languages, these are thought of as _different_ functions from each other.
-(Pattern matching, described just a moment ago, applies only when multiple definitions are provided for function definitions with the _same_ number of arguments.)
+## Local definitions: `where` & `let … in`
 
-### Functions and Pattern Matching
+You can introduce bindings, such as local definition or intermediate results, through two different syntaxes:  
+The `where` keyword, and the `let … in` construct.
```

Comment:
Is this line supposed to be rendered on a new line? It seems that the line breaks early owing to the newline in the file.

Reply:
Yeah the extra `  ` at the end of the previous line is for a newline